presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ExtractSpatialJoins.java

```diff
@@ -331,9 +424,85 @@ else if (alignment < 0) {
+    }
+
+    private static KDBTree loadKdbTree(String tableName, Session session, Metadata metadata, SplitManager splitManager, PageSourceManager pageSourceManager)
+    {
+        QualifiedObjectName name = toQualifiedObjectName(tableName, session.getCatalog().get(), session.getSchema().get());
+        TableHandle tableHandle = metadata.getTableHandle(session, name)
+                .orElseThrow(() -> new PrestoException(INVALID_SPATIAL_PARTITIONING, format("Table not found: %s", name)));
+        Map<String, ColumnHandle> columnHandles = metadata.getColumnHandles(session, tableHandle);
+        List<ColumnHandle> visibleColumnHandles = columnHandles.values().stream()
+                .filter(handle -> !metadata.getColumnMetadata(session, tableHandle, handle).isHidden())
+                .collect(toImmutableList());
+        if (visibleColumnHandles.size() != 1) {
+            throw new PrestoException(INVALID_SPATIAL_PARTITIONING, format("Expected single column for table %s, but found %s columns", name, columnHandles.size()));
+        }
+
+        ColumnHandle kdbTreeColumn = Iterables.getOnlyElement(visibleColumnHandles);
+
+        List<TableLayoutResult> layouts = metadata.getLayouts(session, tableHandle, Constraint.alwaysTrue(), Optional.of(ImmutableSet.of(kdbTreeColumn)));
+        verify(!layouts.isEmpty(), "layouts is empty for table: %s", name);
+
+        Optional<KDBTree> kdbTree = Optional.empty();
+        try (SplitSource splitSource = splitManager.getSplits(session, layouts.get(0).getLayout().getHandle(), UNGROUPED_SCHEDULING)) {
+            while (!Thread.currentThread().isInterrupted()) {
+                SplitBatch splitBatch = getFutureValue(splitSource.getNextBatch(NOT_PARTITIONED, Lifespan.taskWide(), 1000));
+                List<Split> splits = splitBatch.getSplits();
+
+                for (Split split : splits) {
+                    try (ConnectorPageSource pageSource = pageSourceManager.createPageSource(session, split, ImmutableList.of(kdbTreeColumn))) {
+                        do {
+                            getFutureValue(pageSource.isBlocked());
+                            Page page = pageSource.getNextPage();
+                            if (page != null && page.getPositionCount() > 0) {
+                                if (kdbTree.isPresent()) {
+                                    throw new PrestoException(INVALID_SPATIAL_PARTITIONING, format("Expected exactly one row for table %s, but found more", name));
+                                }
+                                if (page.getPositionCount() > 1) {
+                                    throw new PrestoException(INVALID_SPATIAL_PARTITIONING, format("Expected exactly one row for table %s, but found %s rows", name, page.getPositionCount()));
+                                }
+                                kdbTree = Optional.of(KDB_TREE_CODEC.fromJson(VARCHAR.getSlice(page.getBlock(0), 0).toStringUtf8()));
```

Comment:
It may throw `IllegalArgumentException` in case of parsing failure. Catch and categorize.